src/overlays/actors/ovl_Arrow_Ice/z_arrow_ice.c

```diff
@@ -28,33 +33,211 @@ const ActorInit Arrow_Ice_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Arrow_Ice/ArrowIce_Draw.s")
+extern Gfx D_0E0002E0[];
+
+void ArrowIce_SetupAction(ArrowIce* this, ArrowIceActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void ArrowIce_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ArrowIce* this = THIS;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    this->radius = 0;
+    this->unk_158 = 1.0f;
+    ArrowIce_SetupAction(this, ArrowIce_Charge);
+    Actor_SetScale(&this->actor, 0.01f);
+    this->alpha = 100;
+    this->timer = 0;
+    this->unk_15C = 0.0f;
+}
+
+void ArrowIce_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    func_80115D5C(globalCtx);
+    (void)"消滅"; // Unreferenced in retail, means "Disappearance"
+}
+
+void ArrowIce_Charge(ArrowIce* this, GlobalContext* globalCtx) {
+    EnArrow* arrow;
+
+    arrow = (EnArrow*)this->actor.parent;
+    if ((arrow == NULL) || (arrow->actor.update == NULL)) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+
+    if (this->radius < 10) {
+        this->radius += 1;
+    }
+    // copy position and rotation from arrow
+    this->actor.world.pos = arrow->actor.world.pos;
+    this->actor.shape.rot = arrow->actor.shape.rot;
+
+    func_800B9010(&this->actor, NA_SE_PL_ARROW_CHARGE_ICE - SFX_FLAG);
+
+    // if arrow has no parent, player has fired the arrow
+    if (arrow->actor.parent == NULL) {
+        this->unkPos = this->actor.world.pos;
+        this->radius = 10;
+        ArrowIce_SetupAction(this, ArrowIce_Fly);
+        this->alpha = 255;
+    }
+}
+
+void func_809225D0(Vec3f* unkPos, Vec3f* icePos, f32 scale) {
+    unkPos->x += ((icePos->x - unkPos->x) * scale);
+    unkPos->y += ((icePos->y - unkPos->y) * scale);
+    unkPos->z += ((icePos->z - unkPos->z) * scale);
+}
+
+void ArrowIce_Hit(ArrowIce* this, GlobalContext* globalCtx) {
+    f32 scale;
+    f32 offset;
+    u16 timer;
+
+    if (this->actor.projectedW < 50.0f) {
+        scale = 10.0f;
+    } else {
+        if (950.0f < this->actor.projectedW) {
+            scale = 310.0f;
+        } else {
+            scale = this->actor.projectedW;
+            scale = ((scale - 50.0f) * (1.0f / 3.0f)) + 10.0f;
+        }
+    }
+
+    timer = this->timer;
+    if (timer != 0) {
+        this->timer -= 1;
+
+        if (this->timer >= 8) {
+            offset = ((this->timer - 8) * (1.0f / 24.0f));
+            offset = SQ(offset);
+            this->radius = (((1.0f - offset) * scale) + 10.0f);
+            this->unk_158 += ((2.0f - this->unk_158) * 0.1f);
+            if (this->timer < 16) {
+                if (1) {}
+                this->alpha = ((this->timer * 0x23) - 0x118);
+            }
+        }
+    }
+
+    if (this->timer >= 9) {
+        if (this->unk_15C < 1.0f) {
+            this->unk_15C += 0.25f;
+        }
+    } else {
+        if (this->unk_15C > 0.0f) {
+            this->unk_15C -= 0.125f;
+        }
+    }
```

Comment:
Could turn into an else if, although it would lose the parallel.

Reply:
I like the symmetry of it and find it more readable, but if others disagree, I can change it. It's OK as an `else if` though, so it's strictly a matter of opinion